docs/syntax-attributes-functions-parameters.md

```diff
@@ -0,0 +1,34 @@
+# Function Attribute Parameters
+
+## Summary
+
+This RFC proposes a syntax for function attribute parameters. This is a follow up to the [Attributes](./syntax-attributes-functions.md) RFC, which proposed a design for function attributes. We only propose a syntactic extension in this RFC; the rest of the implementation details remain unchanged.
+
+## Motivation
+
+The [Attributes](./syntax-attributes-functions.md) RFC provides a syntax for parameterless function attributes. This suffices for some potential use cases, such as `@native`, and `@inline`. However, other potential use cases such as `@deprecated` will benefit from an expressive attribute syntax that allows parameters to be supplied. We can use the parameters to provide the name of the function to be used in place of the deprecated function, and the reason for deprecation. This information can be used to generate informative deprecation warning messages. Another potential use case could be an `@unroll` attribute for loop unrolling which could use a numeric parameter for specifying the number of iterations to be unrolled. This would also require supporting attributes on loops, which would be a topic of discussion for a separate RFC. It might be desirable to allow these attributes to assume a default value for parameters not provided explicitly, but that's a discussion reserved for attribute-specific RFCs. 
+
+## Design
+
+The following syntax is proposed for attributes with parameters:
+
+```ebnf
+attribute = '@' NAME 
+          | '@' NAME '(' ATOMIC-LITERAL (',' ATOMIC-LITERAL)* ')'
```

Comment:
What happens with this statement? We need some way to delimit the attribute from the `(`.

```lua
@my_cool_attribute
(f()).x = 7
```

Reply:
And no, significant whitespace is _not_ the answer to this.

Reply:
This RFC is implementing attributes for function declarations only. `(f()).x = 7` is not a function declaration. 
If we want the freedom to attach attributes to arbitrary statements, we need another RFC.

Reply:
That's true, but my whole charge is ensuring that the language is forward compatible for the rest of its evolution within the same edition. The syntax as is doesn't allow us this affordability.

Reply:
We know ahead of time if an attribute takes parameters. For instance `@native` does not but `@deprecated` does. The parser will interpret a `(f())` followed by `@deprecated` as an attribute parameter and throw an error since `f()` is not a valid parameter category. 
In general, the ambiguity you mention will only happen if we choose to implement an attribute to have both a parameter and a parameter-less form. In other words, we can avoid this ambiguity, always.

Reply:
I am wondering if I should rework the proposal as follows:

1. Attributes are allowed an optional table parameter (only one parameter which is always a table and nothing else).
2. Attributes without parameters start with a single `@`, like `@native`.
3. Attributes with a parameter start with two `@` symbols, like `@@deprecated <parameter>`.

The use of `@` and `@@` removes the ambiguity you mentioned. The use of a table parameter makes it as general as it could be. I like `@@deprecated {use = "...", reason = "..."}` a lot more over `[deprecated {use = "...", reason = "..."}]`.

Reply:
A simple way to avoid ambiguity, if we choose to support attributes on arbitrary expression, is to introduce empty parameter list for the attribute. For instance, the original example can be written like this:

```
@my_cool_attribute()
(f()).x = 7
```